### PR TITLE
Return RetBuf address via RAX for Windows 64

### PR DESCRIPTION
**Pull request by sejongoh**

### CODE REVIEW

**Comment by RussKeldorph** [comment]
@dotnet-bot test this please

**Comment by sejongoh** [comment]
@sivarv PTAL

File: src/jit/compiler.h [hunk 1]
```diff
@@ -7855,9 +7855,12 @@ public :
         // 2. As per the System V ABI, the address of RetBuf needs to be returned by  
         //    methods with hidden RetBufArg in RAX. In such case GT_RETURN is of TYP_BYREF,  
         //    returning the address of RetBuf.  
-#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+        //
+        // 3. Windows 64-bit native calling convention also requires the address of RetBuff
+        //    to be in RAX.
```
Comment (sivarv) [inline src/jit/compiler.h]: (minor) _"to be returned in RAX."_

File: src/jit/codegenxarch.cpp [hunk 2]
```diff
@@ -209,13 +209,13 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
...
+#ifdef _TARGET_AMD64_
+            // For x64, structs that are not returned in registers are always 
             // returned in implicit RetBuf. If we reached here, we should not have
             // a RetBuf and the return type should not be a struct.
             assert(compiler->info.compRetBuffArg == BAD_VAR_NUM);
             assert(!varTypeIsStruct(compiler->info.compRetNativeType));
-#endif // FEATURE_UNIX_AMD64_STRUCT_PASSING
+#endif // _TARGET_AMD64_
 
             // For Windows we can't make such assertions since we generate code for returning of
```
Comment (sivarv) [inline src/jit/codegenxarch.cpp]: (Minor) _"For x86 Windows"_

File: src/jit/importer.cpp [hunk 3]
```diff
@@ -13904,11 +13904,12 @@ bool Compiler::impReturnInstruction(BasicBlock *block, int prefixFlags, OPCODE &
...
-#if defined(FEATURE_UNIX_AMD64_STRUCT_PASSING)
-        // System V ABI requires to return the implicit return buffer explicitly (in RAX).
+#if defined(_TARGET_AMD64_)
+        // x64 (System V and Win64) calling convention requires to 
+        // return the implicit return buffer explicitly (in RAX).
         // Change the return type to be BYREF.  
         op1 = gtNewOperNode(GT_RETURN, TYP_BYREF, gtNewLclvNode(info.compRetBuffArg, TYP_BYREF));
-#else // defined(FEATURE_UNIX_AMD64_STRUCT_PASSING)
+#else // defined(_TARGET_AMD64_)
         // In case of Windows AMD64 the profiler hook requires to return the implicit RetBuf explicitly (in RAX).
```
Comment (sivarv) [inline src/jit/importer.cpp]: (Minor) this comment needs to be updated?

File: src/jit/importer.cpp [hunk 4]
```diff
@@ -13904,11 +13904,12 @@ bool Compiler::impReturnInstruction(BasicBlock *block, int prefixFlags, OPCODE &
         impAppendTree(op2, (unsigned)CHECK_SPILL_NONE, impCurStmtOffs);
 
         // There are cases where the address of the implicit RetBuf should be returned explicitly (in RAX).  
-#if defined(FEATURE_UNIX_AMD64_STRUCT_PASSING)
-        // System V ABI requires to return the implicit return buffer explicitly (in RAX).
+#if defined(_TARGET_AMD64_)
```
Comment (sivarv) [inline src/jit/importer.cpp]: Were there any asm diffs due to this?

Comment (sivarv) [inline src/jit/importer.cpp]: Please also make sure to update ABI doc that we have in open post this PR merged.

**Comment by RussKeldorph** [comment]
@sejongoh Please add a detailed description.  Also, is this fixing #4529?

**Comment by sejongoh** [comment]
@sivarv  PTAL. I updated the changes per feedback. I saw some asm differences such as 'mov rax, reg' at the end of functions and some different register assignments on the same instructions, but couldn't see any significant and unexpected changes.

File: src/jit/compiler.h [hunk 5]
```diff
@@ -7869,9 +7869,12 @@ public :
...
-#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+        //
+        // 3. Windows 64-bit native calling convention also requires the address of RetBuff
+        //    to be returned in RAX.
+#ifdef _TARGET_AMD64_
         return (info.compRetBuffArg != BAD_VAR_NUM);
-#else // FEATURE_UNIX_AMD64_STRUCT_PASSING  
+#else // _TARGET_AMD64_  
         return (compIsProfilerHookNeeded()) && (info.compRetBuffArg != BAD_VAR_NUM);
 #endif // !FEATURE_UNIX_AMD64_STRUCT_PASSING
```
Comment (sivarv) [inline src/jit/compiler.h]: (Minor) //!_TARGET_AMD64_

File: src/jit/compiler.h [hunk 6]
```diff
@@ -7869,9 +7869,12 @@ public :
...
         //    methods with hidden RetBufArg in RAX. In such case GT_RETURN is of TYP_BYREF,  
         //    returning the address of RetBuf.  
-#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+        //
+        // 3. Windows 64-bit native calling convention also requires the address of RetBuff
+        //    to be returned in RAX.
+#ifdef _TARGET_AMD64_
         return (info.compRetBuffArg != BAD_VAR_NUM);
-#else // FEATURE_UNIX_AMD64_STRUCT_PASSING  
+#else // _TARGET_AMD64_
```
Comment (sivarv) [inline src/jit/compiler.h]: (Minor) //!_TARGET_AMD64_

File: src/jit/importer.cpp [hunk 7]
```diff
@@ -13905,12 +13905,13 @@ bool Compiler::impReturnInstruction(BasicBlock *block, int prefixFlags, OPCODE &
...
+        // x64 (System V and Win64) calling convention requires to 
+        // return the implicit return buffer explicitly (in RAX).
         // Change the return type to be BYREF.  
         op1 = gtNewOperNode(GT_RETURN, TYP_BYREF, gtNewLclvNode(info.compRetBuffArg, TYP_BYREF));
-#else // defined(FEATURE_UNIX_AMD64_STRUCT_PASSING)
-        // In case of Windows AMD64 the profiler hook requires to return the implicit RetBuf explicitly (in RAX).  
+#else // defined(_TARGET_AMD64_)
```
Comment (sivarv) [inline src/jit/importer.cpp]: (Minor) //!_TARGET_AMD64_

Similarly #endif associated with this #else should be changed to
//!_TARGET_AMD64_

**Comment by sivarv** [comment]
Looks good.

**Comment by sejongoh** [comment]
Thanks for review. I update the comments.

**Comment by sejongoh** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test

**Comment by sejongoh** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test

**Comment by sejongoh** [comment]
@dotnet-bot  test OSX x64 Checked Build and Test

**Pull request merged**